src/get-started/write-and-run-unit-tests.md

```diff
@@ -1,10 +1,229 @@
+        num := rand.Uint64()
+        expected := len(strconv.FormatUint(num, 10))
+        actual := StrLenOfUint64Fast(num)
+        require.Equal(t, expected, actual)
+    }
+}
+```
+
+Golang testing detects test functions from `*_test.go` files of the form:
+
+```go
+func TestXxx(*testing.T)
+```
+
+where `Xxx` does not start with a lowercase letter. The function name serves to identify the test routine.
+
+We follow this pattern but using testify assertions instead of out-of-the-box methods, like `Error` or `Fail`, since they are too low level to use.
+
+Almost we use `require.Xxx` for assertions, which imported from `github.com/stretchr/testify/require`. It fails immediately if the assertions fails and we tend to fail tests fast.
+
+Below are the most frequently used assertions:
+
+```go
+func Equal(t TestingT, expected interface{}, actual interface{}, msgAndArgs ...interface{})
+func EqualValues(t TestingT, expected interface{}, actual interface{}, msgAndArgs ...interface{})
+func Len(t TestingT, object interface{}, length int, msgAndArgs ...interface{})
+func NoError(t TestingT, err error, msgAndArgs ...interface{})
+func True(t TestingT, value bool, msgAndArgs ...interface{})
+```
+
+You can find other assertions follow the [documentation](https://pkg.go.dev/github.com/stretchr/testify/require).
+
+### Parallel
+
+Golang testing provides a method of `testing.T` to run tests in parallel:
+
+```go
+t.Parallel()
+```
+
+We leverage this function to run tests as parallel as possible, so that we make full use of the resource accessible.
+
+When some tests should be run in serial, take use of Golang testing [subtests](https://pkg.go.dev/testing#hdr-Subtests_and_Sub_benchmarks) and parallel the parent test only. In this way, tests in the same subtests set run in serial.
+
+```go
+func TestParent(t *testing.T) {
+    t.Parallel()
+    // <setup code>
+    t.Run("Serial 0", func(t *testing.T) { ... })
+    t.Run("Serial 1", func(t *testing.T) { ... })
+    t.Run("Serial 2", func(t *testing.T) { ... })
+    // <tear-down code>
+}
+```
+
+Generally, if a test modifies global configs or failpoints, it should be considered run in serial.
+
+### TestKit
+
+Most of our tests are much more complex than what are described above. For example, to set up a test, we may create a mock storage, a mock session, or even a local database instance.
+
+These functions are known as testkits. Some are used in one package so we implement them in place, others are quite common so we move it to the `testkit` directory.
+
+When you write complex unit tests, you may take a look at what testkits we have now and try to leverage them. If there is none and the case you meet is considered common, add one.
+
+## How to run unit tests?
+
+### Running all tests
+
+You can always run all tests by executing the `gotest` target in Makefile.
+
+```
+make gotest
+```
+
+which is almost equivalent to `go test ./...` but enable and disable failpoints before and after running tests.
```

Comment:
Will you mention how failpoint works?

Reply:
Yes thanks! I should briefly mention it and setup a link to failpoint repo.

Reply:
done.